extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -19,42 +19,157 @@
 import com.hazelcast.jet.impl.serialization.SerializerHookConstants;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.ByteArraySerializer;
 import com.hazelcast.nio.serialization.Serializer;
 import com.hazelcast.nio.serialization.SerializerHook;
 import com.hazelcast.nio.serialization.StreamSerializer;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericArray;
+import org.apache.avro.generic.GenericContainer;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericData.StringType;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryDecoder;
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.DatumReader;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.EncoderFactory;
 import org.apache.avro.util.Utf8;
 
+import javax.annotation.Nonnull;
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
 
 /**
- * Hazelcast serializer hooks for the classes in the {@code
- * com.hazelcast.jet.avrp} package. This is not a public-facing API.
+ * Hazelcast serializer hooks for the classes in the {@code com.hazelcast.jet.avro} package.
  */
-class AvroSerializerHooks {
+public final class AvroSerializerHooks {
+    private AvroSerializerHooks() { }
+
+    public static <T> byte[] serialize(DatumWriter<T> datumWriter, T data) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+            datumWriter.write(data, encoder);
+            encoder.flush();
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericContainerHook implements SerializerHook<GenericContainer> {
+        private final Map<String, Schema> jsonToSchema = new HashMap<>();
```

Comment:
1. these maps have unbounded size. Because we always send a schema (at least in this version) they should be bound in size with some expiration (eg. LRU).
2. if I am not mistaken, `Serializer` must be thread safe, but these maps are not.
3. I am not sure how many times `createSerializer` will be invoked. If only once, then it does make sense to `GenericContainerHook` object around and these maps could be moved to the serializer implementation.

Reply:
1,2. 3b0e1906
3. 4c04f9cc